backends/vulkan/runtime/graph/ops/impl/Reduce.cpp

```diff
@@ -137,15 +155,90 @@ void add_reduce_node(
       resize_reduce_node));
 }
 
+void add_reduce2d_node(
+    ComputeGraph& graph,
+    const ValueRef in,
+    const ValueRef dims_ref,
+    const ValueRef out,
+    const std::string& op_name) {
+  
+  VK_CHECK_COND(
+      !graph.is_buffer_storage(in) && !graph.is_buffer_storage(out),
+      "Vulkan reduction only supports texture storage");
+
+  const int64_t ndim = graph.dim_of(in);
+  
+  // Extract the two dimensions to reduce over
+  const std::vector<int64_t> dims_list = graph.extract_int_or_symint_list(dims_ref);
+  VK_CHECK_COND(dims_list.size() == 2, "reduce2d requires exactly 2 dimensions");
+  
+  int32_t reduce_dim1 = normalize(dims_list[0], ndim);
+  int32_t reduce_dim2 = normalize(dims_list[1], ndim);
+  
+  // Convert to WHCN format
+  reduce_dim1 = nchw_dim_to_whcn_dim(reduce_dim1, ndim);
+  reduce_dim2 = nchw_dim_to_whcn_dim(reduce_dim2, ndim);
+  
+  // Check that the concat dim is not one of the reduction dims
```

Comment:
should also add a check that `graph.packed_dim_of(in)` and `graph.packed_dim_of(out)` is not one of the reduction dims.